x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java

```diff
@@ -61,105 +86,344 @@ public EnrichPolicyResolver(ClusterService clusterService, TransportService tran
+                    if (failure != null) {
+                        failures.add(failure);
                     }
                 }
-            }), LookupResponse::new, threadPool.executor(EsqlPlugin.ESQL_THREAD_POOL_NAME))
+            }
+        }
+        if (targetClusters.size() != policies.size()) {
+            final String reason;
+            if (failures.isEmpty()) {
+                List<String> missingClusters = targetClusters.stream().filter(c -> policies.containsKey(c) == false).sorted().toList();
+                reason = missingPolicyError(policyName, targetClusters, missingClusters, lookupResults);
+            } else {
+                reason = "failed to resolve enrich policy [" + policyName + "]; reason " + failures;
+            }
+            return Tuple.tuple(null, reason);
+        }
+        Map<String, EsField> mappings = new HashMap<>();
+        Collection<String> enrichFields = new LinkedHashSet<>();
+        Map<String, String> concreteIndices = new HashMap<>();
+        ResolvedEnrichPolicy last = null;
+        for (Map.Entry<String, ResolvedEnrichPolicy> e : policies.entrySet()) {
+            ResolvedEnrichPolicy curr = e.getValue();
+            if (last != null && last.matchField().equals(curr.matchField()) == false) {
+                String error = "enrich policy [" + policyName + "] has different match_field across clusters";
+                return Tuple.tuple(null, error);
+            }
+            if (last != null && last.matchType().equals(curr.matchType()) == false) {
+                String error = "enrich policy [" + policyName + "] has different match_type across clusters";
+                return Tuple.tuple(null, error);
+            }
+            // merge mappings
+            for (Map.Entry<String, EsField> m : curr.mapping().entrySet()) {
+                EsField field = m.getValue();
+                field = new EsField(
+                    field.getName(),
+                    EsqlDataTypes.fromTypeName(field.getDataType().typeName()),
+                    field.getProperties(),
+                    field.isAggregatable(),
+                    field.isAlias()
+                );
+                EsField old = mappings.putIfAbsent(m.getKey(), field);
+                if (old != null && old.equals(field) == false) {
+                    String error = "enrich policy [" + policyName + "] has different mapping across clusters";
+                    return Tuple.tuple(null, error);
+                }
+            }
+            // merge enrich fields
+            if (unresolved.mode == Enrich.Mode.ANY
+                && enrichFields.isEmpty() == false
+                && enrichFields.equals(new HashSet<>(curr.enrichFields())) == false) {
+                String error = "enrich policy [" + policyName + "] has different enrich fields across clusters";
+                return Tuple.tuple(null, error);
+            }
+            enrichFields.addAll(curr.enrichFields());
```

Comment:
Apart from the first call to `addAll`, why are the rest necessary? (is this needed for the rest of the Modes?)